src/main/java/cubicchunks/worldgen/generator/custom/biome/replacer/SwampWaterWithLilypadReplacer.java

```diff
@@ -0,0 +1,84 @@
+ *  all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+package cubicchunks.worldgen.generator.custom.biome.replacer;
+
+import cubicchunks.CubicChunks;
+import cubicchunks.api.worldgen.biome.CubicBiome;
+import cubicchunks.world.ICubicWorld;
+import mcp.MethodsReturnNonnullByDefault;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.init.Blocks;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.gen.NoiseGeneratorPerlin;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+
+@ParametersAreNonnullByDefault
+@MethodsReturnNonnullByDefault
+public class SwampWaterWithLilypadReplacer implements IBiomeBlockReplacer {
+
+    private final NoiseGeneratorPerlin noiseGen;
+    private final int seaLevel;
+
+    private SwampWaterWithLilypadReplacer(NoiseGeneratorPerlin noiseGen, int seaLevel) {
+        this.noiseGen = noiseGen;
+        this.seaLevel = seaLevel;
+    }
+
+    @Override public IBlockState getReplacedBlock(IBlockState previousBlock, int x, int y, int z, double dx, double dy, double dz, double density) {
+        if (y == seaLevel - 1 && previousBlock.getBlock() != Blocks.AIR && density + dy < 0) {
+            double noise = noiseGen.getValue(x * 0.25D, z * 0.25D);
+            if (noise > 0) { // if noise < 0 && isSurface
+                return Blocks.WATER.getDefaultState();
+            }
+        }
+        if (y == seaLevel - 1 && previousBlock.getBlock() == Blocks.AIR && density - dy >= 0) {
+            double noise = noiseGen.getValue(x * 0.25D, z * 0.25D);
+            if (noise > 0 && noise < 0.12) {
+                return Blocks.WATERLILY.getDefaultState();
+            }
+        }
+        return previousBlock;
+    }
+
+    public static IBiomeBlockReplacerProvider provider() {
+        return new IBiomeBlockReplacerProvider() {
+
+            private final ResourceLocation OCEAN_LEVEL = CubicChunks.location("water_level");
+
+            @Override public IBiomeBlockReplacer create(ICubicWorld world, CubicBiome biome, BiomeBlockReplacerConfig conf) {
+                return new SwampWaterWithLilypadReplacer(Biome.GRASS_COLOR_NOISE, MathHelper.floor(conf.getDouble(OCEAN_LEVEL)));
+            }
+
+            @Override public Set<ConfigOptionInfo> getPossibleConfigOptions() {
+                return new HashSet<>();
```

Comment:
`Collections.emptySet()` seems to be the most elegant version. I consider it to be a readability improvement.